apm-agent-core/src/main/java/co/elastic/apm/cache/WeakKeySoftValueLoadingCache.java

```diff
@@ -0,0 +1,86 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.cache;
+
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+
+import javax.annotation.Nullable;
+import java.lang.ref.SoftReference;
+
+public class WeakKeySoftValueLoadingCache<K, V> {
+
+    private final WeakConcurrentMap<K, CacheValue<K, V>> cache = new WeakConcurrentMap.WithInlinedExpunction<>();
+    private final ValueSupplier<K, V> valueSupplier;
+
+    public WeakKeySoftValueLoadingCache(ValueSupplier<K, V> valueSupplier) {
+        this.valueSupplier = valueSupplier;
+    }
+
+    @Nullable
+    public V get(K key) {
+        final CacheValue<K, V> cacheValue = cache.get(key);
+        if (cacheValue != null) {
+            return cacheValue.get(key);
+        } else {
+            CacheValue<K, V> value = new CacheValue<>(key, valueSupplier);
+            cache.put(key, value);
+            return value.get(key);
+        }
+    }
+
+    public interface ValueSupplier<K, V> {
+        @Nullable
+        V get(K key);
+    }
+
+    private static class CacheValue<K, V> {
+
+        private final ValueSupplier<K, V> valueSupplier;
+        @Nullable
+        private SoftReference<V> valueReference;
+
+        private CacheValue(K key, ValueSupplier<K, V> valueSupplier) {
+            this.valueSupplier = valueSupplier;
+            V value = valueSupplier.get(key);
+            if (value != null) {
+                valueReference = new SoftReference<>(value);
+            }
+        }
+
+        @Nullable
+        public V get(K key) {
+            if (valueReference == null) {
+                return null;
+            }
+            V value = this.valueReference.get();
+            if (value != null) {
+                return value;
```

Comment:
Logging the reclaimed `SoftReference`s could be useful (`INFO` level - shouldn't be often and if it is we want to know about it ASAP) so we can have an idea of how often they are GCed.
It can be even more useful if compared with the number of keys reclaimed in the underlying map but this may require extend `com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap.WithInlinedExpunction` and override `expungeStaleEntries`...